plugins/api/src/switch_user.cpp

```diff
@@ -0,0 +1,234 @@
+            // We check the proxy user because it covers clients and server-to-server redirects.
+            // Remember, the client and proxy user information is identical until there's a redirect.
+            if (!irods::is_privileged_proxy(*_comm)) {
+                log_api::error(
+                    "Proxy user [{}] does not have permission to switch client user.", _comm->proxyUser.userName);
+                return SYS_PROXYUSER_NO_PRIV;
+            }
+
+            // Return immediately if the client did not provide non-empty strings for the
+            // username and zone.
+            if (const auto ec = check_input(_input); ec != 0) {
+                return ec;
+            }
+
+            const auto* local_zone = getLocalZoneName();
+
+            if (!local_zone) {
+                log_api::error("Could not get name of local zone.");
+                return SYS_INTERNAL_ERR;
+            }
+
+            namespace adm = irods::experimental::administration;
+
+            // This call covers the existence check too.
+            const auto user_type = adm::server::type(*_comm, adm::user{_input->username, _input->zone});
+
+            if (!user_type) {
+                log_api::error("[{}#{}] is not a user in the local zone.", _input->username, _input->zone);
+                return CAT_INVALID_USER;
+            }
+
+            const auto* user_type_string = adm::to_c_str(*user_type);
+            auto& client = _comm->clientUser;
+
+            // Update the user identity associated with the RsComm.
+            std::strncpy(client.userName, _input->username, sizeof(UserInfo::userName));
+            std::strncpy(client.rodsZone, _input->zone, sizeof(UserInfo::rodsZone));
+            std::strncpy(client.userType, user_type_string, sizeof(UserInfo::userType));
+
+            // Set the appropriate privilege level based on whether the user is local or remote
+            // to this zone.
+            if (std::strncmp(_input->zone, local_zone, sizeof(SwitchUserInput::zone)) == 0) {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? LOCAL_PRIV_USER_AUTH : LOCAL_USER_AUTH;
+            }
+            else {
+                client.authInfo.authFlag =
+                    (*user_type == adm::user_type::rodsadmin) ? REMOTE_PRIV_USER_AUTH : REMOTE_USER_AUTH;
+            }
+
+            // iRODS agents do not disconnect from other nodes following a redirect. For long running agents,
+            // this means the API plugin must invoke rc_switch_user() on each connection.
+            //
+            // TODO What happens if this fails half-way through updating the connections?
+            // Should things be rolled back?
+            // Instead, should the server-to-server connections be refreshed instead of calling rc_switch_user()?
+            // Should iRODS agents hold on to connections after redirecting?
+            update_server_to_server_connections(*_comm);
```

Comment:
would this catch any errors from 'inside'?     yes, very interested in the test that triggers this kind of update (and possible failure)...

Reply:
As currently implemented, this does not guard against failures of `rc_switch_user` on connections created by a redirect.

If an error occurs at any point, the connections could be left in an invalid state. This invalid state can be fixed by calling `rc_switch_user` again.

I doubt we have any tests that trigger these cases as they require the client to keep the agents running. I've demonstrated this using the testing environment and a custom client that reuses the same connection.

Reply:
Current plan:
- If dealing with an older server (no `rc_switch_user` available), call `rcDisconnect`
- If dealing with a newer server and an error occurs, call `rcDisconnect`

Reply:
The alternative to optionally disconnecting is to ALWAYS disconnect connections created via a redirect.

This can be thought of as a reset. This will always be the case for older servers. For servers that support this API plugin, disconnecting is pretty wasteful.

Perhaps, the plugin should check the version of the server it is connected to and disconnect if it is dealing with an older server. That would avoid unnecessary network traffic and keep the logs of the remote server clean.

I'm starting to feel like this is the right approach, assuming we can sniff the version of the remote server.

Reply:
Implementation now checks the remote server's version and if it is less than 4.3.1, the plugin calls `rcDisconnect`. Otherwise, `rc_switch_user` is called.

Reply:
This seems correct - no surprises with older servers, and the nice things with the new ones.

I think the only combinations are
- client -> new -> old ... (disconnect)
- client -> new -> new ... (switch user)
- client -> old -> old ... (n/a)
- client -> old -> new (n/a)